File: app01_java_nft/backend/src/main/java/com/gandarych/nftcerts/storage/LocalStubIpfsStorageService.java
package com.gandarych.nftcerts.storage;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.HexFormat;

/**
 * Local, no-network stand-in for {@link IpfsStorageService} used only under the {@code test}
 * profile. Writes content to a local temp directory and returns a deterministic fake
 * {@code ipfs://stub-<sha256-of-content>} URI, so tests never depend on Pinata availability.
 */
@Service
@Profile("test")
public class LocalStubIpfsStorageService implements IpfsStorageService {

    private final ObjectMapper objectMapper;
    private final Path storageDir;

    public LocalStubIpfsStorageService(ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
        try {
            this.storageDir = Files.createTempDirectory("local-stub-ipfs");
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    @Override
    public PinResult pinFile(byte[] content, String filename, String contentType) {
        return storeAndBuildResult(content);
    }

    @Override
    public PinResult pinJson(Object jsonPayload, String name) {
        try {
            byte[] content = objectMapper.writeValueAsBytes(jsonPayload);
            return storeAndBuildResult(content);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    private PinResult storeAndBuildResult(byte[] content) {
        String cid = "stub-" + sha256Hex(content);
        try {
            Files.write(storageDir.resolve(cid), content);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
        return PinResult.of(cid);
    }

    private String sha256Hex(byte[] content) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            return HexFormat.of().formatHex(digest.digest(content));
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException("SHA-256 algorithm not available", e);
        }
    }
}

File: app03_csharp_nft/backend/NftCerts/Storage/LocalStubIpfsStorageService.cs
using System.Text;
using System.Text.Json;
using NftCerts.Hashing;

namespace NftCerts.Storage;

// Offline stand-in for Pinata: writes the pinned bytes to a local temp directory and returns a
// deterministic fake "ipfs://stub-<sha256>" URI. Mirrors app01/app02's LocalStubIpfsStorageService.
public class LocalStubIpfsStorageService : IIpfsStorageService
{
    private readonly Sha256HashingService _hashing = new();
    private readonly string _storageDir = Path.Combine(Path.GetTempPath(), "local-stub-ipfs");

    public LocalStubIpfsStorageService() => Directory.CreateDirectory(_storageDir);

    public PinResult PinFile(byte[] content, string filename, string contentType) => StoreAndBuildResult(content);

    public PinResult PinJson(JsonElement jsonPayload, string name) =>
        StoreAndBuildResult(Encoding.UTF8.GetBytes(jsonPayload.GetRawText()));

    private PinResult StoreAndBuildResult(byte[] content)
    {
        string cid = "stub-" + _hashing.Sha256Hex(content);
        File.WriteAllBytes(Path.Combine(_storageDir, cid), content);
        return PinResult.Of(cid);
    }
}
